### PR TITLE
HBT-157 - Handle Google additional GDPR consent

### PR DESCRIPTION
**Pull request by samiul-shanto**

### CODE REVIEW

**Comment by jbartek25** [comment]
Hey @samiul-shanto,

How did you test this? When I override headerlift.min.js with my local version pointing to a local PBS, I can't see any requests made to any bidders from PBS. Are you aware of this and why this is happening?

**Comment by samiul-shanto** [comment]
Hello @jbartek25 
Sorry for delay reply.

We made a modification to headerlift build. We do following to test:
1. remove `enablePrebidServer` condition check in file [src/prebid.js:228](https://gitlab.azerdev.com/header-bidding/header-bidding-generator/-/blob/master/src/prebid.js#L228) then build headerlift - so all request go to PBS by default, not depend on config
2. Used script redirector for https://hb.improvedigital.com/pbw/headerlift.min.js to our local headerlift build path

 If any request doesn't go to local PBS, then if might be not enable `enablePrebidServer` in config

**Comment by jbartek25** [comment]
That part is clear. What is not clear, I don't see any server-side requests from PBS to Improve /pbs endpoint on HeaderLift sites. Do you?

**Comment by samiul-shanto** [comment]
Do you remember in our last meeting I shown you I got warning message `debug turned off for account` in response when testing with browser. This was the response:

```
{
    "id": "9b193aae-af54-4f09-a55c-1859bdf1499a",
    "ext": {
        "warnings": {
            "general": [
                {
                    "code": 10002,
                    "message": "debug turned off for account"
                }
            ]
        },
        "responsetimemillis": {
            "improvedigital": 266
        },
        "tmaxrequest": 10000,
        "prebid": {
            "auctiontimestamp": 1641209148666
        }
    }
}
```

But when I test with postman, it seems make a request (http://ad.360yield.com/pbs) pbs server.

Here is example Request:

Local PBS: http://localhost:8000/openrtb2/auction

HTTP Body:
```
{
    "id": "1cea8058-fd27-402f-969e-6e01dd558a58",
    "source": {
        "tid": "1cea8058-fd27-402f-969e-6e01dd558a58",
        "ext": {
            "schain": {
                "ver": "1.0",
                "complete": 1,
                "nodes": [
                    {
                        "asi": "headerlift.com",
                        "hp": 1,
                        "sid": "12972"
                    }
                ]
            }
        }
    },
    "tmax": 20000,
    "imp": [
        {
            "id": "ax-skin",
            "ext": {
                "improvedigital": {
                    "placementId": 22592087,
                    "keyValues": {
                        "hb": [
                            "true"
                        ],
                        "pbct": [
                            "2"
                        ]
                    },
                    "bidFloor": 0.0113,
                    "bidFloorCur": "USD"
                }
            },
            "banner": {
                "format": [
                    {
                        "w": 5,
                        "h": 5
                    }
                ]
            }
        }
    ],
    "test": 1,
    "ext": {
        "prebid": {
            "auctiontimestamp": 1638185781607,
            "targeting": {
                "includewinners": true,
                "includebidderkeys": false
            },
            "channel": {
                "name": "pbjs",
                "version": "v5.14.0"
            }
        }
    },
    "cur": [
        "EUR"
    ],
    "site": {
        "publisher": {
            "id": "1"
        },
        "page": "https://deluxemusic.de/"
    },
    "device": {
        "w": 630,
        "h": 980
    },
    "user": {
        "ext": {
            "eids": [
                {
                    "source": "criteo.com",
                    "uids": [
                        {
                            "id": "5Vejl18lMkJNVG9adGg4blcwV3hIWUklMkZWSmY0TzBWRDhhdXVhak5Hck1La1dEemQwYm5DYW41TDNlamVRZHhJdjFKREtKbWhyZlRHRnRNR1lBRVVOOWlMcU5pZE9XckhDN1NmYVElMkZPRFNmOHBLd0RPam83SHVHUkdvaTNQaGh5JTJGJTJCSGNYOTE",
                            "atype": 1
                        }
                    ]
                },
                {
                    "source": "id5-sync.com",
                    "uids": [
                        {
                            "id": "ID5*rGWcLf3suIbuIVL4knDej5pof3cUIBzaTQY-D0mxAdMABgnF4xTwigQ4vchgXkA5AAy_SRruImMkIWEw_7cPwAAScMeS8UmQCr8hg5ySIA8AGJ6KypWholVlRHRgofVSAB7q5w4WsyI-NQM9q3giUQAkRHusKg59gYt9sgrI3HkAKmSOdpKVqPyGs4wd19r8ADDAydWTN1xmbl7hCR64FAA2GbiH916AZcbbHS-JuEYANxgT4wvXYhaVj3djNoNlADvXpoAhiHgadvcBgI0ghgA8McZnWpJCUc7Z4fsAG30AQuEl6WoX2aHlW0V57aspAEg_KvZ_djs90evwLKp4FABOGAVDn2rWKFDpruYgDTAAVNiyiuiK1M6FFr2aK3WmAFnq2h5481LJ_eODelEM5QBa3v6cKS5Bp12IB8VLsv0AYIVGfUciCA_cRzvUZDDLAGbDTvOtIjrtJWPcTJsMbABs6pVcTiRg7oqxBzwzC7EAcu6cX9nRhXytghQ_CMihAHjtBrlW9mCWlhoAwKRvtwB-2iBvhBHaEFSRrmj4OocAhAFrWdu8UiPRWPEPJg5bAIqHeKWUDujygUrdvI5mTw",
                            "atype": 1,
                            "ext": {
                                "linkType": 2
                            }
                        }
                    ]
                },
                {
                    "source": "crwdcntrl.net",
                    "uids": [
                        {
                            "id": "410cfa577bbdce99a40cba1cb6374945a702ea8e971d4c9ec45c52a0e0002d81",
                            "atype": 1
                        }
                    ]
                },
                {
                    "source": "pubcid.org",
                    "uids": [
                        {
                            "id": "f0dc0a55-6ea5-4345-8786-293addaa3a7d",
                            "atype": 1
                        }
                    ]
                }
            ],
            "consent": "CPP6vO1PP6vO1AKAkADEB2CsAP_AAH_AAAYgIXtf_X__bX9j-_59f_t0eY1P9_r3v-QzjhfNt-8F2L_W_L0X42E7NF36pq4KuR4Eu3LBIQNlHMHUTUmwaokVrzHsak2cpyNKJ7LEmnMZO2dYGHtPn9lTuZKY7_78__fz3j-v_t_-39T378X_3_d5_Wu--_O_V399zLv9____39nP___9v-_9-CF4BJhqXkAXZljgybRpVCiBGFYSFQCgAooBhaIrABgcFOysAj1hCwAQmoCMCIEGIKMGAQACCQBIREBIAWCARAEQCAAEAKEBCAAiYBBYAWBgEAAoBoWIEUAQgSEGRwVHKYEBUi0UEtlYAlBXsaYQBlvgRQKL6KjARrNECwMhIWDmOAJAS8WSB5gAAAAA.f_gAAAAAAAAA",
            "ConsentedProvidersSettings": {
                "consented_providers": "1~39.43.46.55.61.66.70.83.89.93.108.117.122.124.131.135.136.143.144.147.149.159.162.167.171.192.196.202.211.218.228.230.239.241.253.259.266.272.286.291.311.317.322.323.326.327.338.367.371.385.389.394.397.407.413.415.424.430.436.440.445.449.453.482.486.491.494.495.501.503.505.522.523.540.550.559.560.568.574.576.584.587.591.733.737.745.780.787.802.803.817.820.821.829.839.864.867.874.899.904.922.931.938.979.981.985.1003.1024.1027.1031.1033.1034.1040.1046.1051.1053.1067.1085.1092.1095.1097.1099.1107.1127.1135.1143.1149.1152.1162.1166.1186.1188.1192.1201.1205.1211.1215.1226.1227.1230.1252.1268.1270.1276.1284.1286.1290.1301.1307.1312.1329.1345.1356.1364.1365.1375.1403.1415.1416.1419.1440.1442.1449.1455.1456.1465.1495.1512.1516.1525.1540.1548.1555.1558.1564.1570.1577.1579.1583.1584.1591.1603.1616.1638.1651.1653.1665.1667.1677.1678.1682.1697.1699.1703.1712.1716.1721.1722.1725.1732.1745.1750.1765.1769.1782.1786.1800.1808.1810.1825.1827.1832.1837.1838.1840.1842.1843.1845.1859.1866.1870.1878.1880.1889.1899.1917.1929.1942.1944.1962.1963.1964.1967.1968.1969.1978.2003.2007.2008.2027.2035.2039.2044.2046.2047.2052.2056.2064.2068.2070.2072.2074.2088.2090.2103.2107.2109.2115.2124.2130.2133.2137.2140.2145.2147.2150.2156.2166.2177.2179.2183.2186.2202.2205.2216.2219.2220.2222.2225.2234.2253.2264.2279.2282.2292.2299.2305.2309.2312.2316.2322.2325.2328.2331.2334.2335.2336.2337.2343.2354.2357.2358.2359.2366.2370.2376.2377.2387.2392.2394.2400.2403.2405.2407.2411.2414.2416.2418.2425.2427.2440.2447.2459.2461.2462.2465.2468.2472.2477.2481.2484.2486.2488.2492.2493.2496.2497.2498.2499.2501.2510.2511.2517.2526.2527.2532.2534.2535.2542.2544.2552.2563.2564.2567.2568.2569.2571.2572.2575.2577.2583.2584.2589.2595.2596.2601.2604.2605.2608.2609.2610.2612.2614.2621.2628.2629.2633.2634.2636.2642.2643.2645.2646.2647.2650.2651.2652.2656.2657.2658.2660.2661.2669.2670.2677.2681.2684.2686.2687.2690.2695.2698.2707.2713.2714.2729.2739.2767.2768.2770.2771.2772.2784.2787.2791.2792.2798.2801.2805.2812.2813.2816.2817.2818.2821.2822.2827.2830.2831.2834.2836.2838.2839.2840.2844.2846.2847.2849.2850.2851.2852.2854.2856.2860.2862.2863.2865.2867.2869.2873.2874.2875.2876.2878.2879.2880.2881.2882.2883.2884.2886.2887.2888.2889.2891.2893.2894.2895.2897.2898.2900.2901.2908.2909.2911.2912.2913.2914.2916.2917.2918.2919.2920.2922.2923.2924.2927.2929.2930.2931.2939.2940.2941.2942.2947.2949.2950.2956.2961.2962.2963.2964.2965.2966.2968.2970.2973.2974.2975.2979.2980.2981.2983.2985.2986.2987.2991.2993.2994.2995.2997.2999.3000.3002.3003.3005.3008.3009.3010.3012.3016.3017.3018.3019.3024.3025.3028.3034.3037.3038.3043.3044.3045.3048.3052.3053.3055.3058.3059.3063.3065.3066.3068.3070.3072.3073.3074.3075.3076.3077.3078.3089.3090.3093.3094.3095.3097.3099.3100.3104.3106.3109.3111.3112.3116.3117.3118.3119.3120.3121.3124.3126.3127.3128.3130.3135.3136.3145.3149.3150.3151.3154.3155.3159.3162.3163.3167.3172.3173.3180.3182.3183.3184.3185.3187.3188.3189.3190.3194.3196.3197.3209.3210.3211.3214.3215.3217.3219.3222.3223.3225.3226.3227.3228.3230.3231.3232.3234.3235.3236.3237.3238.3240.3241.3244.3245.3250.3251.3253.3257.3260.3268.3281"
            }
        }
    },
    "regs": {
        "ext": {
            "gdpr": 1,
            "us_privacy": "1---"
        }
    }
}
```

HTTP Response:
```
{
    "id": "1cea8058-fd27-402f-969e-6e01dd558a58",
    "ext": {
        "debug": {
            "httpcalls": {
                "improvedigital": [
                    {
                        "uri": "http://ad.360yield.com/pbs",
                        "requestbody": "{\"id\":\"1cea8058-fd27-402f-969e-6e01dd558a58\",\"imp\":[{\"id\":\"ax-skin\",\"banner\":{\"format\":[{\"w\":5,\"h\":5}]},\"ext\":{\"bidder\":{\"placementId\":22592087,\"keyValues\":{\"hb\":[\"true\"],\"pbct\":[\"2\"]},\"bidFloor\":0.0113,\"bidFloorCur\":\"USD\"}}}],\"site\":{\"page\":\"https://deluxemusic.de/\",\"publisher\":{\"id\":\"1\"},\"ext\":{\"amp\":0}},\"device\":{\"ua\":\"PostmanRuntime/7.28.4\",\"h\":980,\"w\":630},\"user\":{\"buyeruid\":\"5cd26de0-60ab-4302-8fd2-cad96fe07270\",\"ext\":{\"consent\":\"CPP6vO1PP6vO1AKAkADEB2CsAP_AAH_AAAYgIXtf_X__bX9j-_59f_t0eY1P9_r3v-QzjhfNt-8F2L_W_L0X42E7NF36pq4KuR4Eu3LBIQNlHMHUTUmwaokVrzHsak2cpyNKJ7LEmnMZO2dYGHtPn9lTuZKY7_78__fz3j-v_t_-39T378X_3_d5_Wu--_O_V399zLv9____39nP___9v-_9-CF4BJhqXkAXZljgybRpVCiBGFYSFQCgAooBhaIrABgcFOysAj1hCwAQmoCMCIEGIKMGAQACCQBIREBIAWCARAEQCAAEAKEBCAAiYBBYAWBgEAAoBoWIEUAQgSEGRwVHKYEBUi0UEtlYAlBXsaYQBlvgRQKL6KjARrNECwMhIWDmOAJAS8WSB5gAAAAA.f_gAAAAAAAAA\",\"eids\":[{\"source\":\"criteo.com\",\"uids\":[{\"id\":\"5Vejl18lMkJNVG9adGg4blcwV3hIWUklMkZWSmY0TzBWRDhhdXVhak5Hck1La1dEemQwYm5DYW41TDNlamVRZHhJdjFKREtKbWhyZlRHRnRNR1lBRVVOOWlMcU5pZE9XckhDN1NmYVElMkZPRFNmOHBLd0RPam83SHVHUkdvaTNQaGh5JTJGJTJCSGNYOTE\",\"atype\":1}]},{\"source\":\"id5-sync.com\",\"uids\":[{\"id\":\"ID5*rGWcLf3suIbuIVL4knDej5pof3cUIBzaTQY-D0mxAdMABgnF4xTwigQ4vchgXkA5AAy_SRruImMkIWEw_7cPwAAScMeS8UmQCr8hg5ySIA8AGJ6KypWholVlRHRgofVSAB7q5w4WsyI-NQM9q3giUQAkRHusKg59gYt9sgrI3HkAKmSOdpKVqPyGs4wd19r8ADDAydWTN1xmbl7hCR64FAA2GbiH916AZcbbHS-JuEYANxgT4wvXYhaVj3djNoNlADvXpoAhiHgadvcBgI0ghgA8McZnWpJCUc7Z4fsAG30AQuEl6WoX2aHlW0V57aspAEg_KvZ_djs90evwLKp4FABOGAVDn2rWKFDpruYgDTAAVNiyiuiK1M6FFr2aK3WmAFnq2h5481LJ_eODelEM5QBa3v6cKS5Bp12IB8VLsv0AYIVGfUciCA_cRzvUZDDLAGbDTvOtIjrtJWPcTJsMbABs6pVcTiRg7oqxBzwzC7EAcu6cX9nRhXytghQ_CMihAHjtBrlW9mCWlhoAwKRvtwB-2iBvhBHaEFSRrmj4OocAhAFrWdu8UiPRWPEPJg5bAIqHeKWUDujygUrdvI5mTw\",\"atype\":1,\"ext\":{\"linkType\":2}}]},{\"source\":\"crwdcntrl.net\",\"uids\":[{\"id\":\"410cfa577bbdce99a40cba1cb6374945a702ea8e971d4c9ec45c52a0e0002d81\",\"atype\":1}]},{\"source\":\"pubcid.org\",\"uids\":[{\"id\":\"f0dc0a55-6ea5-4345-8786-293addaa3a7d\",\"atype\":1}]}],\"consented_providers_settings\":{\"consented_providers\":[39,43,46,55,61,66,70,83,89,93,108,117,122,124,131,135,136,143,144,147,149,159,162,167,171,192,196,202,211,218,228,230,239,241,253,259,266,272,286,291,311,317,322,323,326,327,338,367,371,385,389,394,397,407,413,415,424,430,436,440,445,449,453,482,486,491,494,495,501,503,505,522,523,540,550,559,560,568,574,576,584,587,591,733,737,745,780,787,802,803,817,820,821,829,839,864,867,874,899,904,922,931,938,979,981,985,1003,1024,1027,1031,1033,1034,1040,1046,1051,1053,1067,1085,1092,1095,1097,1099,1107,1127,1135,1143,1149,1152,1162,1166,1186,1188,1192,1201,1205,1211,1215,1226,1227,1230,1252,1268,1270,1276,1284,1286,1290,1301,1307,1312,1329,1345,1356,1364,1365,1375,1403,1415,1416,1419,1440,1442,1449,1455,1456,1465,1495,1512,1516,1525,1540,1548,1555,1558,1564,1570,1577,1579,1583,1584,1591,1603,1616,1638,1651,1653,1665,1667,1677,1678,1682,1697,1699,1703,1712,1716,1721,1722,1725,1732,1745,1750,1765,1769,1782,1786,1800,1808,1810,1825,1827,1832,1837,1838,1840,1842,1843,1845,1859,1866,1870,1878,1880,1889,1899,1917,1929,1942,1944,1962,1963,1964,1967,1968,1969,1978,2003,2007,2008,2027,2035,2039,2044,2046,2047,2052,2056,2064,2068,2070,2072,2074,2088,2090,2103,2107,2109,2115,2124,2130,2133,2137,2140,2145,2147,2150,2156,2166,2177,2179,2183,2186,2202,2205,2216,2219,2220,2222,2225,2234,2253,2264,2279,2282,2292,2299,2305,2309,2312,2316,2322,2325,2328,2331,2334,2335,2336,2337,2343,2354,2357,2358,2359,2366,2370,2376,2377,2387,2392,2394,2400,2403,2405,2407,2411,2414,2416,2418,2425,2427,2440,2447,2459,2461,2462,2465,2468,2472,2477,2481,2484,2486,2488,2492,2493,2496,2497,2498,2499,2501,2510,2511,2517,2526,2527,2532,2534,2535,2542,2544,2552,2563,2564,2567,2568,2569,2571,2572,2575,2577,2583,2584,2589,2595,2596,2601,2604,2605,2608,2609,2610,2612,2614,2621,2628,2629,2633,2634,2636,2642,2643,2645,2646,2647,2650,2651,2652,2656,2657,2658,2660,2661,2669,2670,2677,2681,2684,2686,2687,2690,2695,2698,2707,2713,2714,2729,2739,2767,2768,2770,2771,2772,2784,2787,2791,2792,2798,2801,2805,2812,2813,2816,2817,2818,2821,2822,2827,2830,2831,2834,2836,2838,2839,2840,2844,2846,2847,2849,2850,2851,2852,2854,2856,2860,2862,2863,2865,2867,2869,2873,2874,2875,2876,2878,2879,2880,2881,2882,2883,2884,2886,2887,2888,2889,2891,2893,2894,2895,2897,2898,2900,2901,2908,2909,2911,2912,2913,2914,2916,2917,2918,2919,2920,2922,2923,2924,2927,2929,2930,2931,2939,2940,2941,2942,2947,2949,2950,2956,2961,2962,2963,2964,2965,2966,2968,2970,2973,2974,2975,2979,2980,2981,2983,2985,2986,2987,2991,2993,2994,2995,2997,2999,3000,3002,3003,3005,3008,3009,3010,3012,3016,3017,3018,3019,3024,3025,3028,3034,3037,3038,3043,3044,3045,3048,3052,3053,3055,3058,3059,3063,3065,3066,3068,3070,3072,3073,3074,3075,3076,3077,3078,3089,3090,3093,3094,3095,3097,3099,3100,3104,3106,3109,3111,3112,3116,3117,3118,3119,3120,3121,3124,3126,3127,3128,3130,3135,3136,3145,3149,3150,3151,3154,3155,3159,3162,3163,3167,3172,3173,3180,3182,3183,3184,3185,3187,3188,3189,3190,3194,3196,3197,3209,3210,3211,3214,3215,3217,3219,3222,3223,3225,3226,3227,3228,3230,3231,3232,3234,3235,3236,3237,3238,3240,3241,3244,3245,3250,3251,3253,3257,3260,3268,3281]}}},\"test\":1,\"at\":1,\"tmax\":20000,\"cur\":[\"EUR\"],\"source\":{\"tid\":\"1cea8058-fd27-402f-969e-6e01dd558a58\",\"ext\":{\"schain\":{\"ver\":\"1.0\",\"complete\":1,\"nodes\":[{\"asi\":\"headerlift.com\",\"hp\":1,\"sid\":\"12972\"}]}}},\"regs\":{\"ext\":{\"gdpr\":1,\"us_privacy\":\"1---\"}},\"ext\":{\"prebid\":{\"channel\":{\"name\":\"pbjs\",\"version\":\"v5.14.0\"},\"targeting\":{\"pricegranularity\":{\"precision\":2,\"ranges\":[{\"min\":0,\"max\":20,\"increment\":0.1}]},\"includewinners\":true,\"includebidderkeys\":false,\"includebrandcategory\":null,\"includeformat\":false,\"durationrangesec\":null,\"preferdeals\":false}}}}",
                        "requestheaders": {
                            "Content-Type": [
                                "application/json;charset=utf-8"
                            ],
                            "X-Prebid": [
                                "pbs-go/unknown,pbjs/v5.14.0"
                            ]
                        },
                        "responsebody": "",
                        "status": 204
                    }
                ]
            },
            "resolvedrequest": {
                "id": "1cea8058-fd27-402f-969e-6e01dd558a58",
                "imp": [
                    {
                        "id": "ax-skin",
                        "banner": {
                            "format": [
                                {
                                    "w": 5,
                                    "h": 5
                                }
                            ]
                        },
                        "ext": {
                            "improvedigital": {
                                "placementId": 22592087,
                                "keyValues": {
                                    "hb": [
                                        "true"
                                    ],
                                    "pbct": [
                                        "2"
                                    ]
                                },
                                "bidFloor": 0.0113,
                                "bidFloorCur": "USD"
                            }
                        }
                    }
                ],
                "site": {
                    "page": "https://deluxemusic.de/",
                    "publisher": {
                        "id": "1"
                    },
                    "ext": {
                        "amp": 0
                    }
                },
                "device": {
                    "ua": "PostmanRuntime/7.28.4",
                    "h": 980,
                    "w": 630
                },
                "user": {
                    "ext": {
                        "eids": [
                            {
                                "source": "criteo.com",
                                "uids": [
                                    {
                                        "id": "5Vejl18lMkJNVG9adGg4blcwV3hIWUklMkZWSmY0TzBWRDhhdXVhak5Hck1La1dEemQwYm5DYW41TDNlamVRZHhJdjFKREtKbWhyZlRHRnRNR1lBRVVOOWlMcU5pZE9XckhDN1NmYVElMkZPRFNmOHBLd0RPam83SHVHUkdvaTNQaGh5JTJGJTJCSGNYOTE",
                                        "atype": 1
                                    }
                                ]
                            },
                            {
                                "source": "id5-sync.com",
                                "uids": [
                                    {
                                        "id": "ID5*rGWcLf3suIbuIVL4knDej5pof3cUIBzaTQY-D0mxAdMABgnF4xTwigQ4vchgXkA5AAy_SRruImMkIWEw_7cPwAAScMeS8UmQCr8hg5ySIA8AGJ6KypWholVlRHRgofVSAB7q5w4WsyI-NQM9q3giUQAkRHusKg59gYt9sgrI3HkAKmSOdpKVqPyGs4wd19r8ADDAydWTN1xmbl7hCR64FAA2GbiH916AZcbbHS-JuEYANxgT4wvXYhaVj3djNoNlADvXpoAhiHgadvcBgI0ghgA8McZnWpJCUc7Z4fsAG30AQuEl6WoX2aHlW0V57aspAEg_KvZ_djs90evwLKp4FABOGAVDn2rWKFDpruYgDTAAVNiyiuiK1M6FFr2aK3WmAFnq2h5481LJ_eODelEM5QBa3v6cKS5Bp12IB8VLsv0AYIVGfUciCA_cRzvUZDDLAGbDTvOtIjrtJWPcTJsMbABs6pVcTiRg7oqxBzwzC7EAcu6cX9nRhXytghQ_CMihAHjtBrlW9mCWlhoAwKRvtwB-2iBvhBHaEFSRrmj4OocAhAFrWdu8UiPRWPEPJg5bAIqHeKWUDujygUrdvI5mTw",
                                        "atype": 1,
                                        "ext": {
                                            "linkType": 2
                                        }
                                    }
                                ]
                            },
                            {
                                "source": "crwdcntrl.net",
                                "uids": [
                                    {
                                        "id": "410cfa577bbdce99a40cba1cb6374945a702ea8e971d4c9ec45c52a0e0002d81",
                                        "atype": 1
                                    }
                                ]
                            },
                            {
                                "source": "pubcid.org",
                                "uids": [
                                    {
                                        "id": "f0dc0a55-6ea5-4345-8786-293addaa3a7d",
                                        "atype": 1
                                    }
                                ]
                            }
                        ],
                        "consent": "CPP6vO1PP6vO1AKAkADEB2CsAP_AAH_AAAYgIXtf_X__bX9j-_59f_t0eY1P9_r3v-QzjhfNt-8F2L_W_L0X42E7NF36pq4KuR4Eu3LBIQNlHMHUTUmwaokVrzHsak2cpyNKJ7LEmnMZO2dYGHtPn9lTuZKY7_78__fz3j-v_t_-39T378X_3_d5_Wu--_O_V399zLv9____39nP___9v-_9-CF4BJhqXkAXZljgybRpVCiBGFYSFQCgAooBhaIrABgcFOysAj1hCwAQmoCMCIEGIKMGAQACCQBIREBIAWCARAEQCAAEAKEBCAAiYBBYAWBgEAAoBoWIEUAQgSEGRwVHKYEBUi0UEtlYAlBXsaYQBlvgRQKL6KjARrNECwMhIWDmOAJAS8WSB5gAAAAA.f_gAAAAAAAAA",
                        "ConsentedProvidersSettings": {
                            "consented_providers": "1~39.43.46.55.61.66.70.83.89.93.108.117.122.124.131.135.136.143.144.147.149.159.162.167.171.192.196.202.211.218.228.230.239.241.253.259.266.272.286.291.311.317.322.323.326.327.338.367.371.385.389.394.397.407.413.415.424.430.436.440.445.449.453.482.486.491.494.495.501.503.505.522.523.540.550.559.560.568.574.576.584.587.591.733.737.745.780.787.802.803.817.820.821.829.839.864.867.874.899.904.922.931.938.979.981.985.1003.1024.1027.1031.1033.1034.1040.1046.1051.1053.1067.1085.1092.1095.1097.1099.1107.1127.1135.1143.1149.1152.1162.1166.1186.1188.1192.1201.1205.1211.1215.1226.1227.1230.1252.1268.1270.1276.1284.1286.1290.1301.1307.1312.1329.1345.1356.1364.1365.1375.1403.1415.1416.1419.1440.1442.1449.1455.1456.1465.1495.1512.1516.1525.1540.1548.1555.1558.1564.1570.1577.1579.1583.1584.1591.1603.1616.1638.1651.1653.1665.1667.1677.1678.1682.1697.1699.1703.1712.1716.1721.1722.1725.1732.1745.1750.1765.1769.1782.1786.1800.1808.1810.1825.1827.1832.1837.1838.1840.1842.1843.1845.1859.1866.1870.1878.1880.1889.1899.1917.1929.1942.1944.1962.1963.1964.1967.1968.1969.1978.2003.2007.2008.2027.2035.2039.2044.2046.2047.2052.2056.2064.2068.2070.2072.2074.2088.2090.2103.2107.2109.2115.2124.2130.2133.2137.2140.2145.2147.2150.2156.2166.2177.2179.2183.2186.2202.2205.2216.2219.2220.2222.2225.2234.2253.2264.2279.2282.2292.2299.2305.2309.2312.2316.2322.2325.2328.2331.2334.2335.2336.2337.2343.2354.2357.2358.2359.2366.2370.2376.2377.2387.2392.2394.2400.2403.2405.2407.2411.2414.2416.2418.2425.2427.2440.2447.2459.2461.2462.2465.2468.2472.2477.2481.2484.2486.2488.2492.2493.2496.2497.2498.2499.2501.2510.2511.2517.2526.2527.2532.2534.2535.2542.2544.2552.2563.2564.2567.2568.2569.2571.2572.2575.2577.2583.2584.2589.2595.2596.2601.2604.2605.2608.2609.2610.2612.2614.2621.2628.2629.2633.2634.2636.2642.2643.2645.2646.2647.2650.2651.2652.2656.2657.2658.2660.2661.2669.2670.2677.2681.2684.2686.2687.2690.2695.2698.2707.2713.2714.2729.2739.2767.2768.2770.2771.2772.2784.2787.2791.2792.2798.2801.2805.2812.2813.2816.2817.2818.2821.2822.2827.2830.2831.2834.2836.2838.2839.2840.2844.2846.2847.2849.2850.2851.2852.2854.2856.2860.2862.2863.2865.2867.2869.2873.2874.2875.2876.2878.2879.2880.2881.2882.2883.2884.2886.2887.2888.2889.2891.2893.2894.2895.2897.2898.2900.2901.2908.2909.2911.2912.2913.2914.2916.2917.2918.2919.2920.2922.2923.2924.2927.2929.2930.2931.2939.2940.2941.2942.2947.2949.2950.2956.2961.2962.2963.2964.2965.2966.2968.2970.2973.2974.2975.2979.2980.2981.2983.2985.2986.2987.2991.2993.2994.2995.2997.2999.3000.3002.3003.3005.3008.3009.3010.3012.3016.3017.3018.3019.3024.3025.3028.3034.3037.3038.3043.3044.3045.3048.3052.3053.3055.3058.3059.3063.3065.3066.3068.3070.3072.3073.3074.3075.3076.3077.3078.3089.3090.3093.3094.3095.3097.3099.3100.3104.3106.3109.3111.3112.3116.3117.3118.3119.3120.3121.3124.3126.3127.3128.3130.3135.3136.3145.3149.3150.3151.3154.3155.3159.3162.3163.3167.3172.3173.3180.3182.3183.3184.3185.3187.3188.3189.3190.3194.3196.3197.3209.3210.3211.3214.3215.3217.3219.3222.3223.3225.3226.3227.3228.3230.3231.3232.3234.3235.3236.3237.3238.3240.3241.3244.3245.3250.3251.3253.3257.3260.3268.3281"
                        }
                    }
                },
                "test": 1,
                "at": 1,
                "tmax": 20000,
                "cur": [
                    "EUR"
                ],
                "source": {
                    "tid": "1cea8058-fd27-402f-969e-6e01dd558a58",
                    "ext": {
                        "schain": {
                            "ver": "1.0",
                            "complete": 1,
                            "nodes": [
                                {
                                    "asi": "headerlift.com",
                                    "hp": 1,
                                    "sid": "12972"
                                }
                            ]
                        }
                    }
                },
                "regs": {
                    "ext": {
                        "gdpr": 1,
                        "us_privacy": "1---"
                    }
                },
                "ext": {
                    "prebid": {
                        "channel": {
                            "name": "pbjs",
                            "version": "v5.14.0"
                        },
                        "targeting": {
                            "pricegranularity": {
                                "precision": 2,
                                "ranges": [
                                    {
                                        "min": 0,
                                        "max": 20,
                                        "increment": 0.1
                                    }
                                ]
                            },
                            "includewinners": true,
                            "includebidderkeys": false,
                            "includebrandcategory": null,
                            "includeformat": false,
                            "durationrangesec": null,
                            "preferdeals": false
                        }
                    }
                }
            }
        },
        "responsetimemillis": {
            "improvedigital": 3756
        },
        "tmaxrequest": 20000,
        "prebid": {
            "auctiontimestamp": 1641209950806
        }
    }
}
```

**Comment by samiul-shanto** [comment]
While I trying to debug, I found it actually send request to pds (http://ad.360yield.com/pbs) server but got 204 response from server. please take a look at screenshot https://paste.pics/2f00de0953d09d248be808446fd76fee

File: adapters/improvedigital/improvedigital.go [hunk 1]
```diff
@@ -36,11 +57,16 @@ func (a *ImprovedigitalAdapter) MakeRequests(request *openrtb2.BidRequest, reqIn
 
 func (a *ImprovedigitalAdapter) makeRequest(request openrtb2.BidRequest, imp openrtb2.Imp) (*adapters.RequestData, error) {
 	request.Imp = []openrtb2.Imp{imp}
+
 	reqJSON, err := json.Marshal(request)
 	if err != nil {
 		return nil, err
 	}
 
+	if reqCloneJSON, err := a.getJSONWithAdditionalConsent(request, reqJSON); err == nil {
```
Comment (jbartek25) [inline adapters/improvedigital/improvedigital.go]: This feels wrong. You should be manipulating a clone of `request`, and not its JSON form, having to unmarshal it first then marshal it again. 
So please look into how to simply add additional field consented_providers.consented_providers_settings into a cloned `request`, don't touch other fields inside `user.ext`  including `ConsentedProvidersSettings`. Just add the new field `consented_providers`. `user.ext` can in the future contain more fields than you listed in `UserExtReq` and with your current implementation we would lose any new field.

**Comment by samiul-shanto** [comment]
@jbartek25 

Code update as you suggested. Removed UserExtReq, UserExtRes struct and only add consented_providers key.

File: adapters/improvedigital/improvedigital.go [hunk 2]
```diff
@@ -144,3 +148,87 @@ func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType,
...
+		return request
+	}
+	// End request closing
+
+	reqClone.User.Ext = extJson
+
+	return reqClone
+}
+
+func hasAdditionalConsent(consentStr string) (string, error) {
```
Comment (jbartek25) [inline adapters/improvedigital/improvedigital.go]: this function is redundant: it is called only once and the body is just another if-then logic as most of the other code inside applyAdditionalConsentProviders so please remove the function and move the code inside applyAdditionalConsentProviders.

File: adapters/improvedigital/improvedigital.go [hunk 3]
```diff
@@ -144,3 +148,87 @@ func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType,
 		Message: fmt.Sprintf("Failed to find impression for ID: \"%s\"", impID),
 	}
 }
+
+// This method responsible to check/validate/apply additional consent providers
+// If consent provider check/validate/apply failed, this will return original request
+func (a *ImprovedigitalAdapter) applyAdditionalConsentProviders(request openrtb2.BidRequest) openrtb2.BidRequest {
```
Comment (jbartek25) [inline adapters/improvedigital/improvedigital.go]: please rename to addAdditionalConsentProvidersArray

File: adapters/improvedigital/improvedigital.go [hunk 4]
```diff
@@ -144,3 +148,87 @@ func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType,
 		Message: fmt.Sprintf("Failed to find impression for ID: \"%s\"", impID),
 	}
 }
+
+// This method responsible to check/validate/apply additional consent providers
```
Comment (jbartek25) [inline adapters/improvedigital/improvedigital.go]: Actually additional consent provider are already present in the request as a string. This function adds them as an array in user.ext.consented_providers_settings.consented_providers. Please update the comment.

**Review by jbartek25 — CHANGES_REQUESTED** [review]
The code looks good. Just a couple of minor comments.

**Pull request merged**